src/ndb/ndb_internal.c

```diff
@@ -0,0 +1,426 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+#include <string.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <libswiftnav/logging.h>
+#include "libsbp/piksi.h"
+#include "sbp.h"
+#include "ndb.h"
+#include "ndb_internal.h"
+
+#define NDB_THREAD_PRIORITY (LOWPRIO)
+static WORKING_AREA_CCM(ndb_thread_wa, 1756);
+static void ndb_service_thread(void*);
+
+u8 ndb_file_version[MAX_NDB_FILE_VERSION_LEN];
+u8 ndb_file_end_mark = 0xb6;
+
+static bool do_writes_to_flash = false;
+
+static MUTEX_DECL(data_access);
+
+static ndb_element_metadata_t *wq_first = NULL;
+static ndb_element_metadata_t *wq_last = NULL;
+static CONDVAR_DECL(wq_new_data);
+static MUTEX_DECL(wq_access);
+
+static enum ndb_op_code ndb_wq_put(ndb_element_metadata_t* md);
+static enum ndb_op_code ndb_wq_get(ndb_element_metadata_t** md);
+
+static enum ndb_op_code ndb_open_file(ndb_file_t *file);
+static enum ndb_op_code ndb_read(ndb_file_t *f, cfs_offset_t o,
+                                 void *first_element, size_t s);
+
+static enum ndb_op_code do_nv_writes(bool *data_write_ok, bool *md_write_ok);
+static void do_sbp_updates();
+
+enum ndb_op_code ndb_gps_l2cm_l2c_cap_store(u32 *l2c_cap,
+                                            enum ndb_data_source src) NDB_WEAK;
+enum ndb_op_code ndb_gps_l2cm_l2c_cap_read(u32 *l2c_cap) NDB_WEAK;
+void ndb_p3_sbp_update() NDB_WEAK;
+
+
+enum ndb_op_code ndb_wq_put(ndb_element_metadata_t* md)
+{
+  chMtxLock(&wq_access);
+  if (NULL == wq_last) {
+    wq_first = wq_last = md;
+  } else {
+    wq_last->next = md;
+    wq_last = md;
+  }
+  md->next = NULL;
+  chCondSignal(&wq_new_data);
+  chMtxUnlock(&wq_access);
+  return NDB_ERR_NONE;
+}
+
+enum ndb_op_code ndb_wq_get(ndb_element_metadata_t** md)
+{
+  chMtxLock(&wq_access);
+  if (NULL == wq_first) {
+    /* NOTE: chCondwaitTimeout does not care of
+     locking mutex again when timeout occurs*/
+    chCondWaitTimeout(&wq_new_data, NV_WRITE_REQ_TIMEOUT);
```

Comment:
That looks weird to me, but the documentation says:
 * @post    Exiting the function because a timeout does not re-acquire the mutex, the mutex ownership is lost.